src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -596,17 +610,73 @@ public override void Dispose()
 
         private void Dispose(bool disposing)
         {
-            if (_disposed)
+            int disposed = Interlocked.Exchange(ref _disposed, 1);
+            if (disposed != 0)
             {
                 return;
             }
 
-            _disposed = true;
-            _state.Handle.Dispose();
-            Marshal.FreeHGlobal(_state.SendQuicBuffers);
-            if (_stateHandle.IsAllocated) _stateHandle.Free();
-            CleanupSendState(_state);
-            _state.ConnectionState?.RemoveStream(this);
+            bool callShutdown = false;
+            bool abortRead = false;
+            bool releaseHandles = false;
+            lock (_state)
+            {
+                if (_canWrite && _state.ShutdownWriteState == ShutdownWriteState.None)
+                {
+                    callShutdown = true;
+                }
+
+                if (_canRead && _state.ReadState != ReadState.Aborted && _state.ReadState != ReadState.ReadsCompleted)
+                {
+                    abortRead = true;
+                    _state.ReadState = ReadState.Aborted;
+                }
+
+                if (_state.ShutdownState != ShutdownState.Finished)
+                {
+                    _state.ShutdownState = ShutdownState.Pending;
+                }
+                else if (!callShutdown && !abortRead)
+                {
+                    _state.ShutdownState = ShutdownState.Finished;
+                }
+
+                // We already got final event.
+                releaseHandles = Interlocked.Exchange(ref _state.ShutdownDone, 1) == 2;
+            }
+
+            _canRead = false;
+            _canWrite = false;
+
+            if (callShutdown)
+            {
+                try
+                {
+                    // Handle race condition when stream can be closed handling SHUTDOWN_COMPLETE.
+                    StartShutdown(QUIC_STREAM_SHUTDOWN_FLAGS.GRACEFUL, errorCode: 0);
+                } catch (ObjectDisposedException) { };
+            }
+
+            if (abortRead)
+            {
+                try
+                {
+                    StartShutdown(QUIC_STREAM_SHUTDOWN_FLAGS.ABORT_RECEIVE, 0xffffffff);
+                } catch (ObjectDisposedException) { };
+            }
```

Comment:
Don't we need to wait for the stream to complete here, that is, wait for the ShutdownComplete event or similar?

Related: Since that will happen asynchronously, don't we want to implement DisposeAsync so that we don't block when DisposeAsync is called?

Reply:
DisposeAsync currently just call Dispose. But yes, we may fix it but I would do that as separate change. 
As far as the waiting: I don't think so. When I talk about it with @nibank when one resets stream, it gets to undefined state e.g. previous events may or may not happen. When dust settles down we should get ShutdownComplete. 
So I think we do not need to block Dispose but we need to hold to some structures and release them later. (buffers & handles)

Reply:
BTW We can also use IMMEDIATE flag AFAIK but I don't know if that matters.

Reply:
I think you've pinged different person here :smile: